subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformEnvironmentIntegrationTest.groovy

```diff
@@ -0,0 +1,237 @@
+            assert jars.get(1).endsWith("junit-jupiter-engine-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(2).endsWith("junit-jupiter-api-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(3).endsWith("junit-platform-engine-${JUNIT_PLATFORM_VERSION}.jar");
+            assert jars.get(4).endsWith("junit-platform-commons-${JUNIT_PLATFORM_VERSION}.jar");
+            assert jars.get(5).endsWith("junit-jupiter-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(6).endsWith("opentest4j-${OPENTEST4J_VERSION}.jar");
+
+            // And then the distribution-loaded launcher
+            assert jars.get(7).endsWith("junit-platform-launcher-${JUnitCoverage.LATEST_PLATFORM_VERSION}.jar");
+            assert jars.size() == 8;
+        """)
+
+        expect:
+        succeeds "test"
+    }
+
+    def "does not load junit-platform-launcher from distribution when it is on the classpath already"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                implementation 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+
+                runtimeOnly 'org.junit.platform:junit-platform-launcher'
+            }
+        """
+
+        addClasspathTest("""
+            assert jars.get(0).endsWith("junit-jupiter-params-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(1).endsWith("junit-jupiter-engine-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(2).endsWith("junit-jupiter-api-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(3).endsWith("junit-platform-launcher-${JUNIT_PLATFORM_VERSION}.jar");
+            assert jars.get(4).endsWith("junit-platform-engine-${JUNIT_PLATFORM_VERSION}.jar");
+            assert jars.get(5).endsWith("junit-platform-commons-${JUNIT_PLATFORM_VERSION}.jar");
+            assert jars.get(6).endsWith("junit-jupiter-${JUNIT_JUPITER_VERSION}.jar");
+            assert jars.get(7).endsWith("opentest4j-${OPENTEST4J_VERSION}.jar");
+            assert jars.size() == 8;
+        """)
+
+        expect:
+        succeeds "test"
+
+    }
+
+    def "does not load junit-platform-launcher even if it is on the classpath but has a nonstandard-named jar"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                implementation 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+
+                runtimeOnly 'org.junit.platform:junit-platform-launcher'
+            }
+
+            task renameJUnitJars(type: Copy) {
+                from configurations.testRuntimeClasspath
+                into file('build/renamed-classpath')
+                rename { String fileName ->
+                    if (fileName.startsWith('junit')) {
+                        return fileName.replace('junit', 'renamed-junit')
+                    }
+                }
+            }
+
+            testing.suites.test.sources.runtimeClasspath =
```

Comment:
This works, but is there a better way to do this?

Reply:
A detached configuration here would be tedious to deal with since it is not convenient to add dependencies to it. I'm going to leave this for now. I was mostly looking for a way to avoid the `asFileTree.matching { include '**/*.jar' }` here and for a way to set up `renameJUnitJars` to be able to use `renameJUnitJars.outputs.files` directly. Right now, it only returns the `build/renamed-classpath` directly and I would prefer it returns all jars instead of their parent dir.